### PR TITLE
Handle HTTP chunked requests

### LINKED ISSUE #2205: Support Chunked-Encoding for the HTTP server
See https://github.com/actor-framework/actor-framework/blob/a3a63e69844408afe444f0a965ca627674a891c0/libcaf_net/caf/net/http/server.cpp#L212-L217

### PR DESCRIPTION
**Pull request by riemass**
Closes #2205 
Organized chunked requests in server and router. The server will have callbacks for each chunk. 
The router aggregates the chunks and routes them as a normal request. 
Side-quest: added `.devcontainer/build` to .gitignore. Better than accidentally pushing the build directory.

### CODE REVIEW

**Comment by riemass** [comment]
@Neverlord When the router combines the chunks and delegates the request, the header field `Transfer-Encoding: chunked` will still be present, and there will be no `Content-Length` header. Should I leave as-is or maybe modify the headers? I'm more concerned with the absence of the `Content-Length` header.

File: libcaf_net/caf/net/http/router.cpp [hunk 1]
```diff
@@ -101,4 +101,25 @@ ptrdiff_t router::consume(const request_header& hdr, const_byte_span payload) {
...
 
+error router::begin_chunked_message(const net::http::request_header& hdr) {
+  hdr_ = hdr;
+  return error{};
+}
+
+ptrdiff_t router::consume_chunk(const_byte_span body) {
+  CAF_ASSERT(hdr_.valid());
+  body_.insert(body_.end(), body.begin(), body.end());
+  return body.size();
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/router.cpp]:
```suggestion
  return static_cast<ptrdiff_t>(body.size());
```

File: libcaf_net/caf/net/http/server.cpp [hunk 2]
```diff
@@ -185,6 +185,12 @@ class server_impl : public http::server {
             // Transition to the next mode.
             if (hdr_.chunked_transfer_encoding()) {
               mode_ = mode::read_chunks;
+              if (auto err = up_->begin_chunked_message(hdr_); err.valid()) {
+                write_response(status::internal_server_error,
+                               "Failed to initiate chunked message.");
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/server.cpp]:
```suggestion
                               "Failed to initiate chunked transfer.");
```

File: libcaf_net/caf/net/http/server.cpp [hunk 3]
```diff
@@ -220,11 +226,66 @@ class server_impl : public http::server {
...
+            }
+            received_chunks_size_ = 0;
+            mode_ = mode::read_header;
+            return consumed;
+          }
+          if (up_->consume_chunk(remainder.subspan(0, chunk_size)) < 0) {
+            // TODO write response
+            abort(sec::runtime_error, "error while consuming HTTP chunk");
+            return -1;
+          }
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/server.cpp]: Looking at this, maybe `consume_chunk` should also just return `error`. Usually we return the number of bytes consumed but this doesn't really make sense here. Either the upper layer consumes the chunk or not. There's nothing this layer could reasonably do if the upper layer only consumes half of the chunk.

File: libcaf_net/caf/net/http/server.cpp [hunk 4]
```diff
@@ -220,11 +226,66 @@ class server_impl : public http::server {
...
+            write_response(status::payload_too_large,
+                           "Payload exceeds maximum size.");
+            abort(sec::protocol_error, "payload exceeds maximum size");
+            return -1;
+          }
+          if (remainder.size() < chunk_size + 2) {
+            // Configure the policy for the next call to consume.
+            // Await at least chunk line length + chunk_size bytes + crlf.
+            const auto least = input.size() - remainder.size() + chunk_size + 2;
+            const auto most = max_request_size_ - received_chunks_size_;
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/server.cpp]: Isn't this missing the +2 for CLRF?

File: libcaf_net/caf/net/http/server.test.cpp [hunk 5]
```diff
@@ -28,6 +28,7 @@ namespace {
 struct response_t {
   net::http::request_header hdr;
   byte_buffer payload;
+  std::vector<byte_buffer> chunked_payload;
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/server.test.cpp]:
```suggestion
  std::vector<byte_buffer> payload_chunks;
```

File: libcaf_net/caf/net/http/server.test.cpp [hunk 6]
```diff
@@ -282,21 +300,195 @@ SCENARIO("the client sends a multipart HTTP request") {
...
+        check(std::ranges::equal(maybe_res->chunked_payload[0],
+                                 to_const_byte_span("Hello, world!"sv)));
+        check(std::ranges::equal(maybe_res->chunked_payload[1],
+                                 to_const_byte_span("Developer Network"sv)));
+      }
+    }
+  }
+}
+
+SCENARIO("the server handles edge cases for chunked HTTP requests") {
```
Comment (Neverlord) [inline libcaf_net/caf/net/http/server.test.cpp]: This is way to broad and tells me nothing. What's this scenario actually about? Some tests are about sending `bad_request` back to the client on errors while other tests seem unrelated. Probably should be split up into different scenarios.

**Comment by Neverlord** [comment]
> @Neverlord When the router combines the chunks and delegates the request, the header field `Transfer-Encoding: chunked` will still be present, and there will be no `Content-Length` header. Should I leave as-is or maybe modify the headers? I'm more concerned with the absence of the `Content-Length` header.

The header should reflect what the client has communicated. What use case do you have in mind where I would look at the `Content-Length` header field even though I get the payload in a ready-to-use buffer from CAF?

File: libcaf_net/caf/net/http/server.cpp (hunk 3 above)
Comment (riemass) [reply libcaf_net/caf/net/http/server.cpp]: Doesn't, really. I'll refactor it into an error. Maybe the `end_chunked_message` could return the number of consumed bytes, but that info wouldn't really matter to the lower layer.

File: libcaf_net/caf/net/http/server.cpp (hunk 4 above)
Comment (riemass) [reply libcaf_net/caf/net/http/server.cpp]: When I think about it, you're right. The 2 bytes aren't copied to the payload, which means we can receive two more bytes here.

Comment (riemass) [reply libcaf_net/caf/net/http/server.cpp]: This line had a bug where in edge cases the `most` was less that `least` - for failing to account for the line holding the next chunk size.

File: libcaf_net/caf/net/http/server.test.cpp (hunk 6 above)
Comment (riemass) [reply libcaf_net/caf/net/http/server.test.cpp]: Refactored each GIVEN into a separate TEST. The blocks where too small for a scenario IMHO.

**Comment by riemass** [comment]
> The header should reflect what the client has communicated. What use case do you have in mind where I would look at the `Content-Length` header field even though I get the payload in a ready-to-use buffer from CAF?

We're on the same page the. I can't figure a scenario where someone unconditionally reads the `Content-Length` header while having the payload buffer.

**Review by Neverlord — APPROVED** [review]
